File: subsets.ts
function subsets(nums: number[]): number[][] {
    let res: number[][] = []

    let subset: number[] = []
    function dfs(i: number) {
        if (i >= nums.length) {
            res.push([...subset])
            return
        }

        // decision to include nums[i]
        subset.push(nums[i])
        dfs(i + 1)
        // decision to exclude nums[i]
        subset.pop()
        dfs(i + 1)
    }

    dfs(0)
    return res
};

console.log(subsets([1, 2, 3]))
console.log(subsets([0]))
console.log(subsets([]))

File: subsets.cpp
#include <bits/stdc++.h>

using namespace std;
class Solution {
  public:
    vector<vector<int>> subsets(vector<int> &nums) {
        vector<vector<int>> res;
        vector<int> curr;
        dfs(nums, curr, 0, res);

        return res;
    }

  private:
    void dfs(vector<int> nums, vector<int> curr, size_t currIdx,
             vector<vector<int>> &res) {
        if (currIdx >= nums.size()) {
            res.push_back(curr);
            return;
        }

        // include a num
        curr.push_back(nums[currIdx]);
        dfs(nums, curr, currIdx + 1, res);
        // exclude a num
        curr.pop_back();
        dfs(nums, curr, currIdx + 1, res);
    }
};

int main() {
    Solution *sol = new Solution();
    vector<int> nums = {1, 2, 2};
    vector<vector<int>> res;
    res = sol->subsets(nums);

    for (auto it : res) {
        cout << "[";
        for (auto elem : it) {
            cout << elem << ",";
        }
        cout << "], ";
    }
    cout << endl;
    return 0;
}
